soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -659,30 +661,69 @@ async def _main(server):
         if dsp.HIPCIDA & 0x80000000:
             dsp.HIPCIDA = 1<<31 # must ACK any DONE interrupts that arrive!
 
+        if server:
+            # Check if the client connection is alive.
+            if not is_connection_alive(server):
+                lock.acquire()
+                start_output = False
+                lock.release()
+
 class adsp_request_handler(socketserver.BaseRequestHandler):
     """
     The request handler class for control the actions of server.
     """
 
-    def receive_fw(self, filename):
+    def receive_fw(self):
+        header_sz = 80
+
+        log.info("Receiving...")
+        # The first contain the header and the front
+        # data of the firmware
+        d = self.request.recv(BUF_SIZE)
+
+        # Unpacked the header data
+        # Include size(2), filename(42), MD5(32) and pad(2)
+        header = d[:header_sz]
+        total = d[header_sz:]
+        cnt = len(total)
+        s = struct.Struct('I 42s 32s H')
+        unpacked_data = s.unpack(header)
+        log.info(f'unpacked ({unpacked_data})')
+        fsize = unpacked_data[0]
+        fname = unpacked_data[1]
+        md5_tx_b = unpacked_data[2]
+        pad = unpacked_data[3]
+        log.debug(f'pad: {pad}')
+
+        # Receive the remain part of the firmware. We only
+        # receive the specified amount of bytes.
+        while cnt < fsize:
+            data = self.request.recv(BUF_SIZE)
+            if not data:
+                break
+
+            total += data
+            l = len(data)
+            cnt = cnt + l
+
+        log.info(f"Done Receiving {cnt}.")
+
         try:
-            with open(fw_file,'wb') as f:
-                cnt = 0
-                log.info("Receiving...")
-
-                while True:
-                    l = self.request.recv(BUF_SIZE)
-                    ll = len(l)
-                    cnt = cnt + ll
-                    if not l:
-                        break
-                    else:
-                        f.write(l)
+            with open(fname,'wb') as f:
+                f.write(total)
         except Exception as e:
             log.error(f"Get exception {e} during FW transfer.")
-            return 1
+            return None
 
-        log.info(f"Done Receiving {cnt}.")
+        # Check the MD5 of the firmware
+        md5_rx = hashlib.md5(total).hexdigest()
+        md5_tx = md5_tx_b.decode('utf-8')
+
+        if not md5_tx == md5_rx:
```

Comment:
Why not `md5_tx != md5_rx`? Maybe it's just me, but current way is kinda hard to read.